src/storage/admin/RebuildEdgeIndexProcessor.cpp

```diff
@@ -0,0 +1,92 @@
+
+namespace nebula {
+namespace storage {
+
+void RebuildEdgeIndexProcessor::process(const cpp2::RebuildIndexRequest& req) {
+    CHECK_NOTNULL(kvstore_);
+    auto space = req.get_space_id();
+    auto parts = req.get_parts();
+    auto indexID = req.get_index_id();
+    auto itemRet = indexMan_->getEdgeIndex(space, indexID);
+    if (!itemRet.ok()) {
+        cpp2::ResultCode thriftRet;
+        thriftRet.set_code(cpp2::ErrorCode::E_INDEX_NOT_FOUND);
+        codes_.emplace_back(thriftRet);
+        onFinished();
+        return;
+    }
+
+    auto item = itemRet.value();
+    auto edgeType = item->get_schema_id().get_edge_type();
+    LOG(INFO) << "Rebuild Edge Index Space " << space << " Edge Type " << edgeType
+              << " Edge Index " << indexID;
+
+    auto indexPrefix = NebulaKeyUtils::indexPrefix(space, indexID);
+    for (PartitionID part : parts) {
+        // firstly remove the discard index
+        doRemovePrefix(space, part, indexPrefix, true);
+        std::unique_ptr<kvstore::KVIterator> iter;
+        auto prefix = NebulaKeyUtils::prefix(part);
+        auto ret = kvstore_->prefix(space, part, prefix, &iter);
+        if (ret != kvstore::ResultCode::SUCCEEDED) {
+            LOG(ERROR) << "Processing Part " << part << " Failed";
+            this->pushResultCode(to(ret), part);
+            onFinished();
+            return;
+        }
+
+        std::vector<kvstore::KV> data;
+        data.reserve(FLAGS_rebuild_index_batch_size);
+        int32_t batchSize = 0;
+        VertexID currentSrcVertex = -1;
+        VertexID currentDstVertex = -1;
+        while (iter && iter->valid()) {
+            if (batchSize >= FLAGS_rebuild_index_batch_size) {
+                doPut(space, part, std::move(data));
+                data.clear();
+                batchSize = 0;
+            }
+
+            auto key = iter->key();
+            auto val = iter->val();
+
+            if (!NebulaKeyUtils::isEdge(key) ||
+                NebulaKeyUtils::getEdgeType(key) != edgeType) {
+                iter->next();
+                continue;
+            }
+
+            auto source = NebulaKeyUtils::getSrcId(key);
+            auto destination = NebulaKeyUtils::getDstId(key);
+            if (currentSrcVertex == source || currentDstVertex == destination) {
+                iter->next();
+                continue;
+            } else {
+                currentSrcVertex = source;
+                currentDstVertex = destination;
+            }
+            auto ranking = NebulaKeyUtils::getRank(key);
+            auto reader = RowReader::getEdgePropReader(schemaMan_,
+                                                       std::move(val),
+                                                       space,
+                                                       edgeType);
+            auto values = collectIndexValues(reader.get(), item->get_fields());
+            auto indexKey = NebulaKeyUtils::edgeIndexKey(part, indexID, source,
```

Comment:
The operations should be 
1.  remove the old index
2.  insert the new index

And the two operations should be running inside atomicOp interface .